### ISSUE TITLE
SDL_config_minimal.h   typedef redefinition with different types ('unsigned int' vs 'unsigned long')

### ISSUE THREAD

**Issue by courteous** [emscripten-core/emscripten#10320]
I have just upgraded by emscripten from the github and install the latest as per the instalation instructions here 

https://emscripten.org/docs/getting_started/downloads.html
and now I am at the git commit _deaa7e4bfd79246f02f785c8f2174d1e74ab6c00_

after that I tried to compile the project that I am able to compiler successfully  (without the webassembly part ) with gcc 9.2 and Visual Studio.  However when I try to compile with clang i.e. the webaasembly part I get the end the following error:

                                                              
```
In file included from /home/tito/Projects/l88/src/wasm/mainWasm.cpp:41:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten.h:1:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten/emscripten.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL.h:72:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_main.h:25:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_stdinc.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config.h:45:
/home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config_minimal.h:37:22: error: typedef redefinition with different types ('unsigned int' vs 'unsigned long')
typedef unsigned int size_t;
                     ^
/home/tito/Projects/emscripten/emsdk/upstream/lib/clang/10.0.0/include/stddef.h:46:23: note: previous definition is here
typedef __SIZE_TYPE__ size_t;
                      ^
In file included from /home/tito/Projects/l88/src/wasm/mainWasm.cpp:41:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten.h:1:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten/emscripten.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL.h:72:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_main.h:25:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_stdinc.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config.h:45:
/home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config_minimal.h:44:26: error: typedef redefinition with different types ('long long' vs '__int64_t' (aka 'long'))
typedef signed long long int64_t;
                         ^
/usr/include/x86_64-linux-gnu/bits/stdint-intn.h:27:19: note: previous definition is here
typedef __int64_t int64_t;
                  ^
In file included from /home/tito/Projects/l88/src/wasm/mainWasm.cpp:41:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten.h:1:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/emscripten/emscripten.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL.h:72:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_main.h:25:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_stdinc.h:31:
In file included from /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config.h:45:
/home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include/SDL/SDL_config_minimal.h:45:28: error: typedef redefinition with different types ('unsigned long long' vs '__uint64_t' (aka 'unsigned long'))
typedef unsigned long long uint64_t;

```

                  

I was just able to build my project yesterday with the old  emscripten. 

I have found this closed issue here that shows basically the same problem, but unfortunatelly i do not understand what the solution is in this case can someone elaborate on it. 

https://github.com/emscripten-core/emscripten/issues/8569

The whole project seems to compile correctly i.e. clang generates the bitcode and the 
-flto seems to be applied correctly since when i go and find the compiled object in the build directory files i see :

```
file Block.cpp.o
Block.cpp.o: LLVM IR bitcode

```
The compilation is was made on

```
cat /etc/issue.net 
Ubuntu 18.04.3 LT

```

```
./clang++ --version
clang version 10.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 06cfcdcca7de9c88a1e885eff0d0c4c07090ad48)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/tito/Projects/emscripten/emsdk/upstream/bin/.
```

**Comment by sbc100** [comment]
It looks like you are someone mixing your host include files with emscripten's includes.

e.g: `/usr/include/x86_64-linux-gnu/bits/stdint-intn.h:27:19: note: previous definition is here`

Emscripten is cross compiler and should not be using anything in "/usr/include".   

If that isn't enough to fix your issue then can you post the full compiler command line ?

**Comment by courteous** [comment]
Hello Sam thank you for the quick replay.

Indeed when i print the compiler command i see those 3 includes directories that should not be included in particular:
```

-internal-externc-isystem /usr/include/x86_64-linux-gnu 
-internal-externc-isystem /include 
-internal-externc-isystem /usr/include 
```


They are NOT in my CMAKE file but where they are cumming from ?   Google tells me that i need to add the following flag "-nostdinc++" but i would like to understand  how to find out if those directories has been added by CMAKE or by clang i.e. some kind of default. 


here is the full compiler command 

```

/home/tito/Projects/emscripten/emsdk/upstream/bin/clang-10 
-cc1 -triple x86_64-unknown-linux-gnu -emit-llvm-bc -flto -flto-unit -disable-free -main-file-name Block.cpp 
-mrelocation-model pic -pic-level 2 -mthread-model posix -mframe-pointer=all -fmath-errno -fno-rounding-math -masm-verbose -mconstructor-aliases -munwind-tables -target-cpu x86-64 -dwarf-column-info -fno-split-dwarf-inlining -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -v -resource-dir /home/tito/Projects/emscripten/emsdk/upstream/lib/clang/10.0.0 
-I /home/tito/Projects/botan 
-I /home/tito/Projects/botan/build/include 
-I /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include 
-I /home/tito/Projects/emscripten/emsdk/upstream/lib 
-I /usr/local/gcc-9.2/include/c++/9.2.0 
-I /usr/local/gcc-9.2/include/c++/9.2.0/x86_64-linux-gnu 
-I /home/tito/Projects/l88/src/protocol/validator/elements/staticHeader 
-internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0 
-internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/x86_64-linux-gnu/c++/7.4.0 
-internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/x86_64-linux-gnu/c++/7.4.0 
-internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/backward 
-internal-isystem /usr/local/include 
-internal-isystem /home/tito/Projects/emscripten/emsdk/upstream/lib/clang/10.0.0/include 
-internal-externc-isystem /usr/include/x86_64-linux-gnu 
-internal-externc-isystem /include 
-internal-externc-isystem /usr/include 
-W -Wall -Wextra -Werror -Wshadow -Wnon-virtual-dtor -Wunused -pedantic -pedantic-errors -std=gnu++1z -fdeprecated-macro 
-fdebug-compilation-dir /home/tito/Projects/l88/build/src/protocol/validator/elements/staticHeader 
-ferror-limit 19 -fmessage-length 0 -fgnuc-version=4.2.1 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -faddrsig -o CMakeFiles/protocol_validator_elements_static_header.dir/Block.cpp.o -x c++ /home/tito/Projects/l88/src/protocol/validator/elements/staticHeader/Block.cpp
```


and the output of  "clang-10 -v -print-search-dirs"


```

clang version 10.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 06cfcdcca7de9c88a1e885eff0d0c4c07090ad48)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/tito/Projects/emscripten/emsdk/upstream/bin
Found candidate GCC installation: /usr/lib/gcc-cross/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.4.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
programs: =/home/tito/Projects/emscripten/emsdk/upstream/bin:/usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../../x86_64-linux-gnu/bin
libraries: =/home/tito/Projects/emscripten/emsdk/upstream/lib/clang/10.0.0:/usr/lib/gcc/x86_64-linux-gnu/7.4.0:/usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../x86_64-linux-gnu:/lib/x86_64-linux-gnu:/lib/../lib64:/usr/lib/x86_64-linux-gnu:/usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../..:/home/tito/Projects/emscripten/emsdk/upstream/bin/../lib:/lib:/usr/lib

```

**Comment by courteous** [comment]
If i print the include directories from CMAKE like that  i do not see the "-internal-externc-isystem" listed there i.e.

```

get_property(dirs DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
foreach(dir ${dirs})
  message(STATUS "dir='${dir}'")
endforeach()

```



prints out 

```

-- dir='/home/tito/Projects/botan'
-- dir='/home/tito/Projects/botan/build/include'
-- dir='/home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include'
-- dir='/home/tito/Projects/emscripten/emsdk/upstream/lib'
-- dir='/usr/local/gcc-9.2/include/c++/9.2.0'
-- dir='/usr/local/gcc-9.2/include/c++/9.2.0/x86_64-linux-gnu'
```

**Comment by sbc100** [comment]
Are you using the `emcmake` helper script or otherwise specifying the `-DCMAKE_TOOLCHAIN_FILE` yourself (to point to `cmake/Modules/Platform/Emscripten.cmake`)?

**Comment by courteous** [comment]
Hello Sam I am using emconfigure/emmake i.e. 

```
cd build
emconfigure cmake  -DCMAKE_BUILD_TYPE=WASM-Debug ../
emmake make -j16

```
I have completely deleted the build dir and rebuild it and those -internal-externc-isystem includes are still present i.e. when i try to build with clang.  The only thing that is not from the standart ubuntu repository on the build system is the GCC 9.2 which i have build manually with the flag "--disable-multilib", however that is gcc 9.2 and currently the clang from the emscripten repo is pointing to "Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.4.0" which is the standard Ubuntu one.

Do you have any idea what could be the reason for those "-internal-externc-isystem" beeing presented on the search path?

**Comment by sbc100** [comment]
Can you include the full output of that cmake command?

**Comment by courteous** [comment]
not sure what exactlly do you mean my cmake command ? Do you mean the output of the "emconfigure cmake  -DCMAKE_BUILD_TYPE=WASM-Debug ../" ?

**Comment by sbc100** [comment]
Yup.

**Comment by courteous** [comment]
emconfigure cmake  -DCMAKE_BUILD_TYPE=WASM-Debug ../
CMAKE CMAKE_SYSTEM_VERSION is set to     : 1
CMAKE CMAKE_SYSTEM_NAME is set to        : Emscripten
CMAKE CMAKE_GENERATOR is set to          : Unix Makefiles
CMAKE CMAKE_INSTALL_PREFIX is set to     : /usr/local

 
BOTAN_LINK_DIR is                                        : /home/tito/Projects/botan/
BOTAN_INCLUDE_DIR is                                     : /home/tito/Projects/botan/build/include/
GTEST_INCLUDE_DIR is                                     : /home/tito/Projects/l88/src/googletest/googletest/include
GMOCK_INCLUDE_DIR is                                     : /home/tito/Projects/l88/src/googletest/googlemock/include
EMSCRIPTEN_INCLUDE_DIR is                                : /home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include
 
-- Found PythonInterp: /usr/bin/python (found version "2.7.17") 
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
CMAKE_BUILD_TYPE is: WASM-Debug
CMAKE_SYSTEM_NAME is: Emscripten
-- CMAKE_CXX compiler is : Clang
-- USIG Clang
-- CMAKE_CXX compiler is : Clang
-- USIG Clang
-- CMAKE_SYSTEM_NAME now is set to                : Emscripten
-- WASM COMPILER TOOLCHAIN WILL BE EXECUTED
-- Transfering the default index.html and js files to the build directory
-- dir='/home/tito/Projects/botan'
-- dir='/home/tito/Projects/botan/build/include'
-- dir='/home/tito/Projects/emscripten/emsdk/upstream/emscripten/system/include'
-- dir='/home/tito/Projects/emscripten/emsdk/upstream/lib'
-- dir='/usr/local/gcc-9.2/include/c++/9.2.0'
-- dir='/usr/local/gcc-9.2/include/c++/9.2.0/x86_64-linux-gnu'
-- Set make program to src/exe/parallelmake.sh
-- Configuring done
-- Generating done
-- Build files have been written to: /home/tito/Projects/l88/build

**Comment by courteous** [comment]
here is the part that is generating the wasm 

   	set_target_properties(wasmExec PROPERTIES LINK_FLAGS "-std=c++17 -s WASM=1 -s TOTAL_MEMORY=512MB  -s ALLOW_MEMORY_GROWTH=0 -s NO_EXIT_RUNTIME=1  -s VERBOSE=1  --pre-js /home/tito/Projects/l88/src/wasm/preModule.js -s DEMANGLE_SUPPORT=1 -s DISABLE_EXCEPTION_CATCHING=0 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s EXTRA_EXPORTED_RUNTIME_METHODS='['cwrap', 'getValue', 'setValue']'   " )


	target_compile_options(wasmExec PUBLIC ${CMAKE_C_FLAGS} -flto)
	target_compile_options(wasmExec PUBLIC ${CMAKE_C_FLAGS} -emit-llvm)
	target_compile_options(wasmExec PUBLIC ${CMAKE_CXX_FLAGS} -flto)
	target_compile_options(wasmExec PUBLIC ${CMAKE_CXX_FLAGS} -emit-llvm)
   
    file(COPY index.html DESTINATION ${PROJECT_BINARY_DIR}/src/wasm/
    FILE_PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
    NO_SOURCE_PERMISSIONS
    )
    
    file(COPY util.js DESTINATION ${PROJECT_BINARY_DIR}/src/wasm/
    FILE_PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
    NO_SOURCE_PERMISSIONS
    )
    
    MESSAGE(STATUS "Transfering the default index.html and js files to the build directory")

    # =========================================================================================================

    get_property(dirs DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
    foreach(dir ${dirs})
    message(STATUS "dir='${dir}'")
    endforeach()

**Comment by courteous** [comment]
could those includes be coming from the googletest  repo?

**Comment by courteous** [comment]
one more thing that i thing is important here. I was not sure if that could be comming from googlemock/googletest project thus i disabled it completelly in cmake

then deleted the build dir and issued 

`emconfigure cmake  -DCMAKE_BUILD_TYPE=WASM-Debug ../`

I expected to find there in the DependInfo.cmake that file path i.e. "/usr/include/x86_64-linux-gnu "

 If i grep in the build directory for example for one of the include paths i.e. for example 
`grep -R botan .` 
i get bunch of "DependInfo.cmake" files where that include path is found. That is expected since i have explicitelly defined it.

However if i try to search for the path /usr/include/x86_64-linux-gnu i.e:

`grep -R usr |grep -i include |grep -i x86_64 |grep -i linux`

i do not have any single match there. Where are those mistery internal-externc-isystem comming from ?

**Comment by sbc100** [comment]
I would expect to see output from cmake about detecting the compiler:

```
-- The C compiler identification ...
-- Check for working C compiler: ...
```

Maybe you need to a clean run in order for cmake to print all that.

If you can reproduce this in small hello world example that you could share that would be useful.    Those include paths should certainly not be in your INCLUDE_DIRECTORIES

**Comment by courteous** [comment]
Hello Sam 

I still do not know what caused this, but i was able to solve it in a different way. I simply installed a new VM and installed the emscripten as well as python and java and now those 3 additional paths are gone . So i guess i messed up with something although i still do not know what it was. I can not reproduce it in the new VM  since it the new VM those 3 paths are not present. The only thing i can thing about is somehow the manual compilation/installation of gcc 9.2. Anyway during the new clean  installation I faced a new problem related to SIDE_MODULE, but since that is not related to the original issue i decided to cover it in a separate  case.

many thanks for your help
